bot.py

```diff
@@ -0,0 +1,108 @@
+from collections import defaultdict
+from contextlib import contextmanager
+from typing import Text, Dict, List, Generator
+import math
+import requests
+from flask import Flask, request, jsonify
+
+app = Flask(__name__)
+
+inmemory_storage = defaultdict(list)
+
+
+class Conversation(object):
+    def __init__(self, conversation_id: Text, existing_conversation_events: List[Dict]) -> None:
+        """Creates a conversation.
+        Args:
+            existing_conversation_events: Events which happened earlier in this conversation.
+        """
+        self.conversation_id = conversation_id
+        self.existing_conversation_events = existing_conversation_events
+        self.number_existing_conversation_events = len(existing_conversation_events)
+
+    def addd_user_message(self, message: Text) -> None:
+        self.existing_conversation_events.append({"type": "user", "message": message})
+
+    def add_bot_message(self, bot_messages: Text) -> None:
+        self.existing_conversation_events.append({"type": "bot", "message": bot_messages})
+
+    def new_events_dict(self) -> List[Dict]:
+        return self.existing_conversation_events[self.number_existing_conversation_events:]
+
+
+@contextmanager
+def conversationPersistence(conversation_id: Text,) -> Generator[Conversation, None, None]:
+    """Provides conversation history for a certain conversation.
+    Saves any new events to the conversation storage when the context manager is exited.
+
+    Args:
+        conversation_id: The ID of the conversation. This is usually the same as the username.
+
+    Returns:
+        Conversation from the conversation storage.
+    """
+    old_conversation_events = inmemory_storage[conversation_id]
+    # if old_conversation_events is None:
+    # old_conversation_events = []
+    conversation = Conversation(conversation_id, old_conversation_events)
+    yield conversation
+    inmemory_storage[conversation_id] += conversation.new_events_dict()
+
+
+class ChuckNorrisBot:
+    def handle_message(self, message_text: Text, conversation: Conversation) -> None:
+        conversation.addd_user_message(message_text)
+
+        if len(conversation.existing_conversation_events) <= 1:
+            conversation.add_bot_message(f"Welcome! Let me tell you a joke.")
+
+        joke = self.retrieve_joke()
+        conversation.add_bot_message(joke)
+
+    def retrieve_joke(self) -> Text:
+        response = requests.get("https://api.chucknorris.io/jokes/random")
+
+        return response.json()["value"]
+
+
+@app.route("/user/<username>/message", methods=["POST"])
+def handle_user_message(username: Text) -> Text:
+    """Returns a bot response for an incoming user message.
+    Args:
+        username: The username which serves as unique conversation ID.
+
+    Returns:
+        The bot's responses.
+    """
+    message_text = request.json["text"]
+
+    f = ChuckNorrisBot()
```

Comment:
Ideally we should rename this. `f.handle_message` doesn't mean anything to anyone else reading this